### PR TITLE
WIP: Improvement in the kubeconfigwriter

### PR DESCRIPTION
**Pull request by Divyansh42**
# Changes
`Kubeconfigwriter` needs to improved to replace pipeline resource type "cluster" with the` kubeconfig-creator `task.
In order to do that, some extra configurations (e.g. `client-key-data`, `client-certificate-data`) needs to be added with some changes in the implementation.

Ref: [https://github.com/tektoncd/catalog/issues/95], [https://github.com/tektoncd/catalog/pull/251]

Signed-off-by: Divyansh42 <diagrawa@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```
Added two more configurations in the Resource struct, those are:
- ClientKeyData
- ClientCertificateData

This change will allow user to access the cluster without using token or password,  provided they will have to use ClientKeyData and ClientCertificateData.
```

### CODE REVIEW

File: cmd/kubeconfigwriter/main.go [hunk 1]
```diff
@@ -23,16 +23,48 @@ import (
...
+	resource "github.com/tektoncd/pipeline/pkg/apis/resource/v1alpha1"
 	"github.com/tektoncd/pipeline/pkg/logging"
 	"go.uber.org/zap"
 	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
 	"k8s.io/client-go/tools/clientcmd"
 	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 )
+// Resource represents a cluster configuration (kubeconfig)
+// that can be accessed by tasks in the pipeline
+type Resource struct {
```
Comment (nikhil-thomas) [inline cmd/kubeconfigwriter/main.go:35]: shall we reuse the existing `cluster.Resource` struct
https://github.com/tektoncd/pipeline/blob/a8946d42a91808b96047a1a3090237e787d5d395/pkg/apis/resource/v1alpha1/cluster/cluster_resource.go#L34

File: cmd/kubeconfigwriter/main.go [hunk 2]
```diff
@@ -95,7 +131,7 @@ func createKubeconfigFile(resource *cluster.Resource, logger *zap.SugaredLogger)
 	c.APIVersion = "v1"
 	c.Kind = "Config"
 
-	destinationFile := fmt.Sprintf("/workspace/%s/kubeconfig", resource.Name)
+	destinationFile := fmt.Sprintf("%s/kubeconfig", *destinationDir)
```
Comment (nikhil-thomas) [inline cmd/kubeconfigwriter/main.go:134]: will this change break the existing `cluster` `pipelineResource`

**Comment by vdemeester** [comment]
/ok-to-test

@Divyansh42 is there test covering this ? :upside_down_face:

**Comment by bobcatfish** [comment]
Thanks for improving this @Divyansh42 ! @vdemeester pointed out that we'll want tests for this, a few more things we'll want as well:

- Updates to our docs (e.g. https://github.com/tektoncd/pipeline/blob/master/docs/resources.md#cluster-resource)
- Unit tests, maybe an end to end test also if it makes sense (i think this is the existing test https://github.com/tektoncd/pipeline/blob/d6e21bc705b58edd070135d18fff7bc127667325/test/cluster_resource_test.go)

Thanks again! :D

**Comment by Divyansh42** [comment]
> /ok-to-test
> 
> @Divyansh42 is there test covering this ?

I guess no :sweat_smile: , will update as soon as possible. 
Thanks for the review :)

**Comment by Divyansh42** [comment]
> Thanks for improving this @Divyansh42 ! @vdemeester pointed out that we'll want tests for this, a few more things we'll want as well:
> 
> * Updates to our docs (e.g. https://github.com/tektoncd/pipeline/blob/master/docs/resources.md#cluster-resource)
> * Unit tests, maybe an end to end test also if it makes sense (i think this is the existing test https://github.com/tektoncd/pipeline/blob/d6e21bc705b58edd070135d18fff7bc127667325/test/cluster_resource_test.go)
> * Can you fill in the "release notes" section in the pull request description as well?
> 
> Thanks again! :D

Thanks for the review, I will try to update all these and commit as soon as possible :)

**Pull request closed without merging**